refactored define_alleles and calculate_frequencies and added tests for them

JKimG1 opened the pull request.

jkim3157 committed:
refactored define_alleles and calculate_frequencies and added tests

JHDeerin commented on tests/test_calculate_frequencies.py:
```diff
@@ -0,0 +1,344 @@
+    AlgorithmInstance._get_samples_from_site(additional, 'Zaire'),'Additional_')
+
+@pytest.fixture
+def Benguela_define_alleles():
+    define_alleles = []
+    columns = ['0', '1']
+    index_0 = pd.DataFrame(
+        [[219.0, 221.0],[221.0, 223.0],[243.0, 245.0],[261.0, 263.0],[223.0, 225.0],
+         [225.0, 227.0],[231.0, 233.0],[237.0, 239.0],[239.0, 241.0],[249.0, 251.0],
+         [217.0, 219.0],[229.0, 231.0],[245.0, 247.0]],
```
Holy cow, copying all of these must've taken forever; mad respect. Thanks for doing this!

JHDeerin commented on api/define_alleles.py:
```diff
@@ -5,7 +5,6 @@
 def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np.ndarray):
 	'''
 	Generate definitions of alleles (i.e. binning)
```
Could you explain briefly what the allele definitions are/what binning means?

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -117,19 +64,85 @@ def calculate_frequencies3(genotypedata, alleles_definitions):
+
+	return ret
+
+def _getLocinames(genotypedata: pd.DataFrame):
```
Both calculate_frequencies and define_alleles use this function; can you move it into `recrudescence_utils` and import it into both files instead?

JHDeerin replied:
Also, add a brief docstring/method comment to this explaining what it does (not everyone knows what locinames are, so explain that)

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -117,19 +64,85 @@ def calculate_frequencies3(genotypedata, alleles_definitions):
+				locinames[lociname_index] = (prev_pos, lociname_end_index)
+	return locinames
+
+def _getRawAlleles(genotypedata: pd.DataFrame, n: int, last_index: int):
```
What is `n` here? Could you give it a clearer name?

JHDeerin replied:
Also, add a small docstring explaining what this is doing/returning

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -117,19 +64,85 @@ def calculate_frequencies3(genotypedata, alleles_definitions):
+	raw_alleles = [loci for loci in raw_alleles if str(loci) != 'nan']
+	return raw_alleles, n
+
+def _get_sumList(nrows: int, raw_alleles: list, low: pd.core.series.Series, high: pd.core.series.Series):
```
Add a brief docstring explaining what this is doing/returning

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -117,19 +64,85 @@ def calculate_frequencies3(genotypedata, alleles_definitions):
+
+	return sum_list, meanSD
+
+def _create_frequencyMatrix(nloci: int, ncol: np.int64, frequencies: list):
```
Add a brief docstring explaining what this is doing/returning (what's the frequency matrix?)

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -117,19 +64,85 @@ def calculate_frequencies3(genotypedata, alleles_definitions):
-	# ret = list
-	# add freq_length, frequency matrix, and variability
-	# ret needed to be list type to add each numpy array and matrix
+def _pack_result(freq_length: np.ndarray, freqmatrix: np.ndarray, variability: np.ndarray):
```
This'll lead to changes in other files, but: instead of putting all of these in a list, could you instead put them in a [named tuple](https://pymotw.com/2/collections/namedtuple.html), or a small class w/ each array as a public variable?

Since these 3 matrices don't really fit together, I think that makes more sense than the weird list stuff that Mat did. Calling `frequencies_RR.lengths` would be a lot clearer IMO than `frequencies_RR[0]` everywhere

JHDeerin replied:
Also: docstring (I know these are helper methods, but something really brief like "Puts the 3 frequency-related matrices together into a single result class" helps a ton when reading through this later and understanding why this method exists)

JHDeerin commented on api/define_alleles.py:
```diff
@@ -43,52 +54,22 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+def _getAlleles(genotypedata: pd.DataFrame, nloci: int, locirepeats: np.ndarray, locinames: dict) -> list:
 	alleles = []
 	n = 0
 	for j in range(nloci):
```
Use `i` instead of `j` (since it's the outermost loop in this function)

JHDeerin commented on api/define_alleles.py:
```diff
@@ -43,52 +54,22 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
-	nloci = len(locinames)
-
-	# first section
+def _getAlleles(genotypedata: pd.DataFrame, nloci: int, locirepeats: np.ndarray, locinames: dict) -> list:
```
Docstring?

JHDeerin commented on api/define_alleles.py:
```diff
@@ -43,52 +54,22 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+			lower_break_value, upper_break_value, counts_column = _findBreakValues(raw_alleles, locirepeats, j)
+			alleles = _prepareDataframes(alleles, lower_break_value, upper_break_value, counts)
 		else:
 			# making breaks (not sure if we need this)
```
This else statement is still really huge; could you break it out into its own function, since it seems to be doing something different from the if clause above?

JHDeerin commented on api/define_alleles.py:
```diff
@@ -171,23 +151,50 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+			alleles = _prepareDataframes(alleles, lower_break_value, upper_break_value, counts)
+	return alleles
+
+def _findBreakValues(raw_alleles: list, locirepeats: np.ndarray, j: int):
```
Docstring? Also, use `i` instead of `j`

JHDeerin commented on api/define_alleles.py:
```diff
@@ -171,23 +151,50 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+	counts_column.append(len(lower_list))
+	return lower_break_value, upper_break_value, counts_column
+
+def _getRawAlleles(genotypedata: pd.DataFrame, n: int, last_index: int):
```
Docstring? And what does `n` mean (is there a clearer name)?

JHDeerin commented on api/define_alleles.py:
```diff
@@ -171,23 +151,50 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+	raw_alleles = [loci for loci in raw_alleles if str(loci) != 'nan']
+	return raw_alleles, n
+
+def _prepareDataframes(alleles: list, lower_break_value: list, upper_break_value: list, counts: list):
```
Docstring?

JHDeerin commented on api/define_alleles.py:
```diff
@@ -171,23 +151,50 @@ def define_alleles(genotypedata: pd.DataFrame, locirepeats: np.ndarray, maxk: np
+	alleles.append(allele_df)
+	return alleles
+
+def _compress(alleles: list, nloci: int, maxk: int) -> list:
```
Docstring?

JHDeerin commented on api/calculate_frequencies.py:
```diff
@@ -28,50 +28,21 @@
 """
 def calculate_frequencies3(genotypedata, alleles_definitions):
 
```
Move the comment above inside the function, as a docstring (per standard python format); minor formatting things, but for the comment use single quotes instead of double quotes, and use the `:param (...): (...)`/`:return: (...)` format used by the other methods instead of input/output as it currently uses (just so we're using consistent comment formats across the project)

JHDeerin replied:
Also, could you elaborate a little bit on what "calculate frequencies of alleles" means?

JHDeerin commented on tests/test_define_alleles.py:
```diff
@@ -0,0 +1,222 @@
+        pd.concat([genotypedata_RR_Benguela, additional_neutral_Benguela]), locirepeats, maxk
+    )
+
+    # assert result_list.length == expected_Benguela_define_alleles.length
```
Delete this comment if it's not being used?

JHDeerin commented on tests/test_define_alleles.py:
```diff
@@ -0,0 +1,222 @@
+        pd.concat([genotypedata_RR_Lunda_Sul, additional_neutral_Lunda_Sul]), locirepeats, maxk
+    )
+
+    # assert result_list.length == expected_Benguela_define_alleles.length
```
Delete this comment

JHDeerin commented on tests/test_define_alleles.py:
```diff
@@ -0,0 +1,222 @@
+        pd.concat([genotypedata_RR_Zaire, additional_neutral_Zaire]), locirepeats, maxk
+    )
+
+    # assert result_list.length == expected_Benguela_define_alleles.length
```
Delete this comment

JHDeerin reviewed (changes requested):
While I made a lot of small requests (especially to make docstrings for helper methods, even if they're really short 1-liners and don't talk about each input parameter), the overall structure of this looks solid and the massive majority of the work is already done. The tests especially should give us a lot of peace that this is working like the R code, so a HUGE thanks for doing that!

jkim3157 committed:
added docstrings and refactored more

JHDeerin reviewed (approved):
Looks good!

Again, I think changing `pack_result`/`calculate_frequencies` to return a class w/ the matrices as named variables, instead of a list, makes sense (and'd make the code readable in the long term) and I saw you haven't resolved that yet. If you can do that before you merge, great - please do so!

If not, though, the rest of the changes look good and it should be in a mergeable state.

jkim3157 committed:
added the class Frequencies in calculate_frequencies.py

Added the class that holds the frequency related values. Other files that had frequencies_RR[0] (or [1] or [2]) are now adjusted to frequencies_RR.lengths (or .matrix or .variability)

JHDeerin reviewed (approved):
Looks great! Thanks for doing this! Closes #25 and closes #21

JHDeerin committed:
Merge branch 'master' of https://github.com/GTJuniorDesign0100-2020/anti-malarial-MCMC-bayesian-algorithm into refactor_defAlleles_calFrequencies
